Refactor LSTM tests

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39920 Enable dynamic quantized GRU support
* #37159 Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell
* **#38851 Refactor LSTM tests**
Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

raghuramank10000 committed:
Refactor LSTM tests



Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"


Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"


Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"


Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"


Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

supriyar commented on test/quantization/test_quantized_module.py:
```diff
@@ -728,3 +728,48 @@ def test_linear_api(self, batch_size, in_features, out_features, use_bias, use_d
+        # Check that module matches the numerics of the op and ensure that module can be
+        # instantiated for all engines and dtypes
+
+        for qengine in supported_qengines:
```
You can use the @override_qengines decorator instead of these two lines.

supriyar commented on test/quantization/test_quantized_module.py:
```diff
@@ -728,3 +728,48 @@ def test_linear_api(self, batch_size, in_features, out_features, use_bias, use_d
+                                                              bidirectional=bidirectional,
+                                                              dtype=dtype)
+                    _all_params = ([m.param for m in cell_dq._all_weight_values])
+                    result = torch.quantized_lstm(x, (h, c), _all_params, cell_dq.bias, cell_dq.num_layers,
```
Do we also have a test for the quantize_dynamic api for LSTM modules? Should we test that as well?

supriyar reviewed (commented):
Should we also add a test at op level for quantized lstm? That tests for different input shapes and layers? I don't see one in test_quantized_op.py

supriyar commented on test/quantization/test_quantize.py:
```diff
@@ -447,8 +445,6 @@ def checkQuantized(model):
         checkQuantized(model_oneline)
 
-
-@skipIfNoFBGEMM
```
Why remove this decorator? It will fail in environments without fbgemm support.

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank100 commented:
> Should we also add a test at op level for quantized lstm? That tests for different input shapes and layers? I don't see one in test_quantized_op.py

I am adding this in PR: #37159

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

z-a-f reviewed (approved):
lgtm

supriyar commented on test/quantization/test_quantized_op.py:
```diff
@@ -2128,6 +2130,129 @@ def test_qlinear(self, batch_size, input_channels, output_channels,
+                if dtype == torch.qint8:
+                    packed_ih = torch.ops.quantized.linear_prepack(Wq1, b1)
+                    packed_hh = torch.ops.quantized.linear_prepack(Wq2, b2)
+                    cell_params = torch.ops.quantized.make_quantized_cell_params_dynamic(packed_ih, packed_hh, b1, b2)
```
We should set `reduce_range` True here.

supriyar commented on test/quantization/test_quantize.py:
```diff
@@ -680,59 +678,15 @@ def checkQuantized(model):
-
-        ref_out, ref_hid = ref(x, hiddens)
-
         for qengine in supported_qengines:
```
Can we use `@override_qengines` decorator here?

supriyar commented on test/quantization/test_quantize.py:
```diff
@@ -680,59 +678,15 @@ def checkQuantized(model):
 
     def test_quantized_rnn(self):
-        r"""Test execution and serialization for dynamic quantized lstm modules on int8 and fp16
+        r"""Test dynamic quantization, scriptability and serialization for dynamic quantized lstm modules on int8 and fp16
```
There is a test below `test_per_channel_lstm_quantize` for per-channel quantization. Should we combine the two here and go over both qconfigs?

raghuramank100 replied:
Yes, I will combine the tests

supriyar commented on test/quantization/test_quantized_op.py:
```diff
@@ -2128,6 +2130,129 @@ def test_qlinear(self, batch_size, input_channels, output_channels,
+        def _get_rnn_inputs(seq_len, num_batches, input_size, hidden_size, num_directions):
+            # For Input (seq_len, batch, input_size)
+            X = torch.randn(seq_len, num_batches, input_size)
+            # TODO: Change to reduce_range=True once support is enabled
```
This is supported now

raghuramank100 replied:
I have it enabled in the next PR in the stack, will move the change to this PR itself.

supriyar commented on test/quantization/test_quantized_op.py:
```diff
@@ -2128,6 +2130,129 @@ def test_qlinear(self, batch_size, input_channels, output_channels,
+                H = torch.zeros(num_directions, num_batches, hidden_size)
+                C = torch.zeros(num_directions, num_batches, hidden_size)
+
+            s, z = _calculate_dynamic_qparams(H, torch.quint8, reduce_range=False)
```
Change `reduce_range` to true here

supriyar commented on test/quantization/test_quantized_op.py:
```diff
@@ -2128,6 +2130,129 @@ def test_qlinear(self, batch_size, input_channels, output_channels,
+
+            s, z = _calculate_dynamic_qparams(H, torch.quint8, reduce_range=False)
+            Hq = torch.quantize_per_tensor(H, s, z, torch.quint8)
+            s, z = _calculate_dynamic_qparams(C, torch.quint8, reduce_range=False)
```
Same here

raghuramank100 replied:
Thanks, have made all the changes requested.

supriyar reviewed (changes requested):
Left a few comments that I think should be addressed before submitting.

z-a-f reviewed (approved):
Testing the differential approval

z-a-f reviewed (approved):
Testing phabricator

z-a-f reviewed (approved):
Testing the phabricator

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

z-a-f reviewed (approved):
Testing phabricator

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Refactor LSTM tests"

Current LSTM tests do not check for scriptability. We also do not test that the module is consistent with the LSTM op. Modified the tests and refactored them to fit into the new testing structure
Differential Revision: [D21628596](https://our.internmc.facebook.com/intern/diff/D21628596/)

[ghstack-poisoned]
